File: 강의 자료/02-알고리즘/15~17-동적 프로그래밍/문제별 코드/1495-기타리스트/solution.py
import sys
si = sys.stdin.readline

n, S, M = list(map(int, si().split()))
a = [0] + list(map(int, si().split()))
dy = [[False for __ in range(M + 1)] for _ in range(n + 1)]

# 초기값 구하기
dy[0][S] = True

# 점화식을 토대로 dy 배열 채우기
for i in range(1, n + 1):
    ans = -1
    for prev in range(M + 1):
        if not dy[i - 1][prev]: continue
        for cur in [prev - a[i], prev + a[i]]:
            if cur < 0 or cur > M: continue
            ans = max(ans, cur)
            dy[i][cur] = True

print(ans)

File: 강의 자료/02-알고리즘/15~17-동적 프로그래밍/문제별 코드/1495-기타리스트/solution.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#pragma warning(disable:4996)
using namespace std;
#define NM 55

int N, S, M;
int A[NM];
bool Dy[NM][1001];

void input() {
	cin >> N >> S >> M;
	for (int i = 1; i <= N; i++) cin >> A[i];
}

void pro() {
	// 초기값 구하기
	Dy[0][S] = true;

	// 점화식을 토대로 Dy 배열 채우기
	int ans = 0;
	for (int i = 1; i <= N; i++) {
		bool flag = false;
		ans = 0;
		for (int prev = 0; prev <= M; prev++) {
			if (!Dy[i - 1][prev]) continue;
			for (int cur : {prev - A[i], prev + A[i]}) {
				if (cur < 0 || cur > M) continue;
				ans = max(ans, cur);
				Dy[i][cur] = true;
				flag = true;
			}
		}
		if (!flag) {
			cout << -1;
			return;
		}
	}

	// Dy배열로 정답 계산하기
	cout << ans;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	input();
	pro();
	return 0;
}